pkg/kubelet/kubelet_pods.go

```diff
@@ -1001,6 +1002,67 @@ func (kl *Kubelet) removeOrphanedPodStatuses(pods []*v1.Pod, mirrorPods []*v1.Po
 	kl.statusManager.RemoveOrphanedStatuses(podUIDs)
 }
 
+// deleteOrphanedMirrorPods checks orphaned mirror pod whether its grace period has elapsed.
+// Once the grace period passes, podManager.DeleteMirrorPod() is called to delete mirror pod
+// from the API server
+func (kl *Kubelet) deleteOrphanedMirrorPods() {
+	podFullNames := kl.podManager.GetOrphanedMirrorPodNames()
+	start := time.Now()
+	for _, podFullName := range podFullNames {
+		uidWithGracePeriod, ok := kl.podManager.GetPendingPodUID(podFullName)
+		if !ok {
+			klog.Warningf("unable to find pod UID for %q", podFullName)
+			continue
+		}
+		allContainersExited := true
+		if start.Before(uidWithGracePeriod.Till) {
+			name, namespace, err := kubecontainer.ParsePodFullName(podFullName)
+			if err != nil {
+				klog.Warningf("unable to parse %q : %v", podFullName, err)
+				continue
+			}
+			podStatus, err := kl.containerRuntime.GetPodStatus(types.UID(uidWithGracePeriod.UID), name, namespace)
+			if err != nil {
+				klog.Errorf("unable to retrieve pod status for %q / %q : %v", namespace, name, err)
+				continue
+			}
+			if len(podStatus.ContainerStatuses) == 0 {
+				klog.V(3).Infof("there is no container status for pod %q / %q", namespace, name)
+				continue
+			}
+			// check that all the containers have exited
+			for _, containerStatus := range podStatus.ContainerStatuses {
+				if containerStatus.State != kubecontainer.ContainerStateExited {
+					klog.V(3).Infof("container %q for %q has state %v", containerStatus.Name, name, containerStatus.State)
+					allContainersExited = false
+					break
+				}
+			}
+			if allContainersExited {
+				klog.V(3).Infof("all containers have exited for %q / %q (%v)", namespace, name, uidWithGracePeriod.UID)
+			}
+		} else {
+			klog.V(3).Infof("%q has passed the grace period %v", podFullName, uidWithGracePeriod.Till)
+		}
+		// the following condition, when true, covers two scenarios:
+		// 1. grace period has passed. the if block for start.Before() is not executed.
+		// 2. within grace period, CRI tells us that all the containers have exited
+		if allContainersExited {
+			deleted, err := kl.podManager.DeleteMirrorPod(podFullName, nil)
+			if err != nil {
+				klog.Errorf("encountered error when deleting mirror pod %q : %v", podFullName, err)
+				continue
+			}
+			if !deleted {
+				klog.Errorf("mirror pod %q was not deleted", podFullName)
+				continue
+			}
+			// the mirror pod has been removed from api server, clean the entry for bookkeeping
+			kl.podManager.DeletePendingPod(podFullName)
```

Comment:
i see what looks like a leak on this map as it has entries for all pods, but we only remove mirror pods.

did i miss something here?